rir/src/compiler/util/visitor.h

```diff
@@ -12,26 +12,83 @@
 namespace rir {
 namespace pir {
 
-class Visitor {
+namespace VisitorHelpers {
+typedef std::function<bool(BB*)> BBActionPredicate;
+typedef std::function<void(BB*)> BBAction;
+
+/*
+ * PredicateWrapper abstracts over BBAction and BBActionPredicate, to be
+ * able to use them in the same generic implementation.
+ * In the case of BBAction the return value will always be true.
+ *
+ */
+template <typename ActionKind>
+struct PredicateWrapper {};
+
+template <>
+struct PredicateWrapper<BBAction> {
+    const BBAction action;
+    bool operator()(BB* bb) const {
+        action(bb);
+        return true;
+    }
+};
+
+template <>
+struct PredicateWrapper<BBActionPredicate> {
+    const BBActionPredicate action;
+    bool operator()(BB* bb) const { return action(bb); }
+};
+
+/*
+ * Helpers to remember which BB has already been visited. There is a fast
```

Comment:
Would be useful to point out that IDMarker is the fast version and PointerMarker is the slow version.